Counter

ALeksey5678 opened the pull request.

ALeksey5678 committed:
working program

ALeksey5678 committed:
working program and Tests

ALeksey5678 committed:
working program and Tests for revision

saber0008 commented on src/main/java/com/aleksey5678/counter_string_modifier/CachingUniqueCharactersCounter.java:
```diff
@@ -0,0 +1,31 @@
+package com.aleksey5678.counter_string_modifier;
```
please don't use `_` in package names

saber0008 replied:
judging by this package name we definitely should split it in two packages

ALeksey5678 replied:
Done

saber0008 commented on src/main/java/com/aleksey5678/counter_string_modifier/UniqueCharactersCounter.java:
```diff
@@ -0,0 +1,26 @@
+    private static final int ONE = 1;
+    private static final int ZERO = 0;
+
+    public Map<Character, Integer> showUniqueSymbolsAndTheirQuantity(String ourSentence) {
```
I think `show*` method should not return map.
Name could be something like `calculate*`, `summarize*`, `count*` or something

ALeksey5678 replied:
Done

saber0008 commented on src/main/java/com/aleksey5678/counter_string_modifier/UniqueCharactersCounter.java:
```diff
@@ -0,0 +1,26 @@
+
+    public Map<Character, Integer> showUniqueSymbolsAndTheirQuantity(String ourSentence) {
+
+       Map<Character, Integer> storageOfUniqueSymbolsAndTheirQuantity = new HashMap<>();
```
I suggest to reflect map contents in name. Please look at this thread https://stackoverflow.com/questions/2253453/how-should-i-name-a-java-util-map there is some good options for map namings

ALeksey5678 replied:
Done

saber0008 commented on src/main/java/com/aleksey5678/cache/Cache.java:
```diff
@@ -0,0 +1,23 @@
+import java.util.Map;
+
+public class Cache {
+  private final Map<String, String> cacheStorage;
```
please reflect content or map in name. See details my other comment

ALeksey5678 replied:
Done

saber0008 commented on src/main/java/com/aleksey5678/cache/Cache.java:
```diff
@@ -0,0 +1,23 @@
+        this.cacheStorage = cacheStorage;
+    }
+
+    public boolean isKeyHere(String ourSentence) {
```
can be `isCached`

ALeksey5678 replied:
Done

saber0008 commented on src/main/java/com/aleksey5678/cache/Cache.java:
```diff
@@ -0,0 +1,23 @@
+
+import java.util.Map;
+
+public class Cache {
```
Too generic name for this cache

ALeksey5678 replied:
Done

saber0008 commented on src/main/java/com/aleksey5678/cache/Cache.java:
```diff
@@ -0,0 +1,23 @@
+        return cacheStorage.get(ourSentence);
+    }
+
+    public void putInCache(String ourSentence, String resultOfCalculations) {
```
can be just `save`

ALeksey5678 replied:
Done

saber0008 commented on src/main/java/com/aleksey5678/counter_string_modifier/CachingUniqueCharactersCounter.java:
```diff
@@ -0,0 +1,31 @@
+        this.cache = cache;
+    }
+
+    public String countUniqueCharactersOrGetFromCache(String ourSentence) {
```
I suggest not to use `our*` in any names, we are building the application not the communism :) 
Same for `my*`, `your*`

ALeksey5678 replied:
Done

saber0008 commented on src/main/java/com/aleksey5678/counter_string_modifier/Modifier.java:
```diff
@@ -0,0 +1,12 @@
+
+import java.util.Map;
+
+public class Modifier {
```
too generic

ALeksey5678 replied:
Done

saber0008 commented on src/main/java/com/aleksey5678/counter_string_modifier/Modifier.java:
```diff
@@ -0,0 +1,12 @@
+    public String modify(Map<Character, Integer> quantityOfUniqueSymbols) {
+        StringBuilder stringBuilder = new StringBuilder();
+        quantityOfUniqueSymbols.forEach((character, quantity) ->
+                stringBuilder.append(String.format("%s - %s",character,quantity)).append(System.lineSeparator()));
```
please extract formatting pattern to constant

ALeksey5678 replied:
Done

saber0008 commented on src/main/java/com/aleksey5678/counter_string_modifier/UniqueCharactersCounter.java:
```diff
@@ -0,0 +1,26 @@
+
+       Map<Character, Integer> storageOfUniqueSymbolsAndTheirQuantity = new HashMap<>();
+
+        for (int indexOfCharacter = ZERO; indexOfCharacter < ourSentence.length(); indexOfCharacter++) {
```
since we are not using indexes here we can just use for each loop, or `forEach()` method

ALeksey5678 replied:
Done

saber0008 commented on src/main/java/com/aleksey5678/counter_string_modifier/UniqueCharactersCounter.java:
```diff
@@ -0,0 +1,26 @@
+        for (int indexOfCharacter = ZERO; indexOfCharacter < ourSentence.length(); indexOfCharacter++) {
+            char characters = ourSentence.charAt(indexOfCharacter);
+            if (!storageOfUniqueSymbolsAndTheirQuantity.containsKey(characters)) {
+                storageOfUniqueSymbolsAndTheirQuantity.put(characters, ONE);
```
I suggest to use computeIfAbsent() here

ALeksey5678 replied:
Done

saber0008 commented on src/main/java/com/aleksey5678/counter_string_modifier/UniqueCharactersCounter.java:
```diff
@@ -0,0 +1,26 @@
+            if (!storageOfUniqueSymbolsAndTheirQuantity.containsKey(characters)) {
+                storageOfUniqueSymbolsAndTheirQuantity.put(characters, ONE);
+            } else {
+                Integer quantityOfUniqueSymbols = storageOfUniqueSymbolsAndTheirQuantity.get(characters);
```
I suggest to use computeIfPresent() here

ALeksey5678 replied:
Done

ALeksey5678 committed:
working program and Tests for revision(after 1 pull request)

saber0008 commented on src/test/java/com/aleksey5678/uniquecharacterscounter/cache/CalculationResultCacheTest.java:
```diff
@@ -0,0 +1,50 @@
+package com.aleksey5678.uniquecharacterscounter.cache;
```
packages in tests and prod code should be same

saber0008 commented on src/main/java/com/aleksey5678/characterscounter/characterscounter/UniqueCharactersCounter.java:
```diff
@@ -0,0 +1,18 @@
+
+        Map<Character, Integer> characterToInteger = new HashMap<>();
+        for (char characters : sentence.toCharArray()) {
+            characterToInteger.computeIfPresent(characters, (character, integer) -> integer+ONE);
```
formatting

ALeksey5678 replied:
Done

saber0008 commented on src/test/java/com/aleksey5678/uniquecharacterscounter/cache/CalculationResultCacheTest.java:
```diff
@@ -0,0 +1,50 @@
+package com.aleksey5678.uniquecharacterscounter.cache;
```
packages in tests and prod code should be same

ALeksey5678 replied:
Done

saber0008 commented on src/main/java/com/aleksey5678/characterscounter/cache/CalculationResultCache.java:
```diff
@@ -0,0 +1,23 @@
+        this.stringMap = stringMap;
+    }
+
+    public boolean isCached(String ourSentence) {
```
our

ALeksey5678 replied:
Done

saber0008 commented on src/main/java/com/aleksey5678/characterscounter/characterscounter/CachingUniqueCharactersCounter.java:
```diff
@@ -0,0 +1,32 @@
+    private final CalculationResultCache calculationResultCache;
+
+    public CachingUniqueCharactersCounter
+            (UniqueCharactersCounter uniqueCharactersCounter, MapToStringModifier mapToStringModifier, CalculationResultCache calculationResultCache) {
```
formatting

ALeksey5678 replied:
Done

saber0008 replied:
not exactly what I mean
please don't put opening bracket on new line, instead you can break line after comma
should be `CachingUniqueCharactersCounter(*`

ALeksey5678 replied:
Done

saber0008 commented on src/main/java/com/aleksey5678/characterscounter/stringmodifier/MapToStringModifier.java:
```diff
@@ -0,0 +1,14 @@
+
+import java.util.Map;
+
+public class MapToStringModifier {
```
imo `MapToStringFormatter` will be better

ALeksey5678 replied:
Done

saber0008 commented on src/main/java/com/aleksey5678/characterscounter/stringmodifier/MapToStringModifier.java:
```diff
@@ -0,0 +1,14 @@
+public class MapToStringModifier {
+    private static final String FORMATTING_PATTERN = "%s - %s";
+
+    public String modify(Map<Character, Integer> quantityOfUniqueSymbols) {
```
this is not `modify` operation

saber0008 replied:
also please please update name of parameter

ALeksey5678 replied:
Done

saber0008 commented on src/test/java/com/aleksey5678/uniquecharacterscounter/cache/CalculationResultCacheTest.java:
```diff
@@ -0,0 +1,50 @@
+    private Map<String, String> storageMock;
+
+    @InjectMocks
+    CalculationResultCache newCalculationResultCache;
```
please always incapsulate fields

saber0008 replied:
don't need `new` in name

ALeksey5678 replied:
Done

saber0008 commented on src/test/java/com/aleksey5678/uniquecharacterscounter/cache/CalculationResultCacheTest.java:
```diff
@@ -0,0 +1,50 @@
+
+    @InjectMocks
+    CalculationResultCache newCalculationResultCache;
+    private final String sentence = "a";
```
could be constant

ALeksey5678 replied:
Done

saber0008 commented on src/test/java/com/aleksey5678/uniquecharacterscounter/cache/CalculationResultCacheTest.java:
```diff
@@ -0,0 +1,50 @@
+    @InjectMocks
+    CalculationResultCache newCalculationResultCache;
+    private final String sentence = "a";
+    private final String resultOfCalculations = "a - 1";
```
constant

saber0008 replied:
also please add `expected*` to name

ALeksey5678 replied:
Done

saber0008 commented on src/main/java/com/aleksey5678/characterscounter/characterscounter/UniqueCharactersCounter.java:
```diff
@@ -0,0 +1,18 @@
+    public Map<Character, Integer> calculateUniqueSymbolsAndTheirQuantity(String sentence) {
+
+        Map<Character, Integer> characterToInteger = new HashMap<>();
+        for (char characters : sentence.toCharArray()) {
```
I suggest to keep same terminology
since we as using word `symbol` in method name we can also use it here instead of `character`.

ALeksey5678 replied:
Done

saber0008 commented on src/test/java/com/aleksey5678/uniquecharacterscounter/cache/CalculationResultCacheTest.java:
```diff
@@ -0,0 +1,50 @@
+@ExtendWith(MockitoExtension.class)
+class CalculationResultCacheTest {
+    @Mock
+    private Map<String, String> storageMock;
```
also please dont forget to update name of map here

ALeksey5678 replied:
Done

saber0008 commented on src/test/java/com/aleksey5678/uniquecharacterscounter/cache/CalculationResultCacheTest.java:
```diff
@@ -0,0 +1,50 @@
+    void shouldReturnTrueIfCached() {
+        when(storageMock.containsKey(sentence)).thenReturn(true);
+
+        boolean cached = newCalculationResultCache.isCached(sentence);
```
please use `is*` prefix with boolean variables

ALeksey5678 replied:
Done

ALeksey5678 committed:
working program and Tests for revision(after 1 pull request)
Map names  changed

saber0008 commented on src/test/java/com/aleksey5678/uniquecharacterscounter/cache/CalculationResultCacheTest.java:
```diff
@@ -0,0 +1,50 @@
+    @Test
+    void shouldPutKeyAndValueInCache() {
+        newCalculationResultCache.save(sentence, resultOfCalculations);
+        verify(storageMock).put(sentence, resultOfCalculations);
```
please always use `verify()` with times. In this case `verify(storageMock, times(1)).*`

ALeksey5678 replied:
Done

saber0008 commented on src/test/java/com/aleksey5678/uniquecharacterscounter/characterscounter/CachingUniqueCharactersCounterTest.java:
```diff
@@ -0,0 +1,59 @@
+    private CalculationResultCache calculationResultCacheMock;
+
+    @InjectMocks
+   private CachingUniqueCharactersCounter cachingUniqueCharactersCounter;
```
formatting

ALeksey5678 replied:
Done

saber0008 commented on src/test/java/com/aleksey5678/uniquecharacterscounter/characterscounter/CachingUniqueCharactersCounterTest.java:
```diff
@@ -0,0 +1,59 @@
+@ExtendWith(MockitoExtension.class)
+class CachingUniqueCharactersCounterTest {
+    @Mock
+    private UniqueCharactersCounter uniqueCharactersCounterMock;
```
no need for `*mock` suffix in names

ALeksey5678 replied:
Done

saber0008 commented on src/test/java/com/aleksey5678/uniquecharacterscounter/characterscounter/CachingUniqueCharactersCounterTest.java:
```diff
@@ -0,0 +1,59 @@
+    @InjectMocks
+   private CachingUniqueCharactersCounter cachingUniqueCharactersCounter;
+
+    private final String sentence = "a";
```
constant

ALeksey5678 replied:
Done

saber0008 commented on src/test/java/com/aleksey5678/uniquecharacterscounter/stringmodifier/MapToStringModifierTest.java:
```diff
@@ -0,0 +1,26 @@
+class MapToStringModifierTest {
+
+    @Test
+    void shouldModifyValue() {
```
what is `value`?

ALeksey5678 replied:
Done

saber0008 commented on src/test/java/com/aleksey5678/uniquecharacterscounter/stringmodifier/MapToStringModifierTest.java:
```diff
@@ -0,0 +1,26 @@
+    void shouldModifyValue() {
+        String expectedResult = "a - 1";
+
+        Map<Character, Integer> map = new HashMap<>();
```
generic name

ALeksey5678 replied:
Done

saber0008 commented on src/test/java/com/aleksey5678/uniquecharacterscounter/stringmodifier/MapToStringModifierTest.java:
```diff
@@ -0,0 +1,26 @@
+
+        Map<Character, Integer> map = new HashMap<>();
+        map.put('a', 1);
+        MapToStringModifier mapToStringModifier = new MapToStringModifier();
```
please move it to field

ALeksey5678 replied:
Done

saber0008 commented on src/test/java/com/aleksey5678/uniquecharacterscounter/stringmodifier/MapToStringModifierTest.java:
```diff
@@ -0,0 +1,26 @@
+
+import static org.junit.jupiter.api.Assertions.assertEquals;
+
+class MapToStringModifierTest {
```
please also add test for `null` parameter

ALeksey5678 replied:
Done

saber0008 commented on src/test/java/com/aleksey5678/uniquecharacterscounter/characterscounter/UniqueCharactersCounterTest.java:
```diff
@@ -0,0 +1,26 @@
+
+import static org.junit.jupiter.api.Assertions.assertEquals;
+
+class UniqueCharactersCounterTest {
```
please also add test for null parameter

ALeksey5678 replied:
Done

ALeksey5678 committed:
many changes+tests if parameter is Null(last time making commits like that,sorry)

saber0008 commented on src/main/java/com/aleksey5678/uniquecharacterscounter/characterscounter/UniqueCharactersCounter.java:
```diff
@@ -0,0 +1,25 @@
+
+        Map<Character, Integer> numberOfCharacterRepetitionsByUniqueCharacter = new HashMap<>();
+
+        for (char characters : sentence.toCharArray()) {
```
should be `character`. Variable of type `char` can only represent single character

ALeksey5678 replied:
Done

saber0008 commented on src/test/java/com/aleksey5678/uniquecharacterscounter/characterscounter/CachingUniqueCharactersCounterTest.java:
```diff
@@ -0,0 +1,57 @@
+        String result = cachingUniqueCharactersCounter.countUniqueCharactersOrGetFromCache(SENTENCE);
+
+        assertEquals(EXPECTED_RESULT, result);
+        verify(calculationResultCache).save(SENTENCE, result);
```
please always use `times(*)` with `verify`

ALeksey5678 replied:
Done

saber0008 commented on src/test/java/com/aleksey5678/uniquecharacterscounter/characterscounter/CachingUniqueCharactersCounterTest.java:
```diff
@@ -0,0 +1,57 @@
+
+        assertEquals(EXPECTED_RESULT, result);
+        verify(calculationResultCache).save(SENTENCE, result);
+        verify(uniqueCharactersCounter).calculateUniqueCharactersAndTheirQuantity(SENTENCE);
```
`times`

ALeksey5678 replied:
Done

saber0008 commented on src/test/java/com/aleksey5678/uniquecharacterscounter/stringformatter/MapToStringFormatterTest.java:
```diff
@@ -0,0 +1,34 @@
+    }
+
+    @Test
+    void returnNullPointerExceptionIfParameterIsNull() {
```
this kind of tests should be like `shouldThrow*`.
return is not very correct word for exceptions, exceptions are thrown not returned

ALeksey5678 replied:
Done

saber0008 commented on src/test/java/com/aleksey5678/uniquecharacterscounter/stringformatter/MapToStringFormatterTest.java:
```diff
@@ -0,0 +1,34 @@
+    private final MapToStringFormatter mapToStringFormatter = new MapToStringFormatter();
+
+    @Test
+    void shouldModifyQuantityOfSymbolRepetitions() {
```
`shouldFormatMapAsString`.

ALeksey5678 replied:
Done

saber0008 commented on src/test/java/com/aleksey5678/uniquecharacterscounter/characterscounter/UniqueCharactersCounterTest.java:
```diff
@@ -0,0 +1,37 @@
+    }
+
+    @Test
+    void shouldReturnNullPointerExceptionIfParameterIsNull() {
```
this kind of tests should be like shouldThrow*.
return is not very correct word for exceptions, exceptions are thrown not returned

ALeksey5678 replied:
Done

saber0008 commented on src/main/java/com/aleksey5678/uniquecharacterscounter/characterscounter/UniqueCharactersCounter.java:
```diff
@@ -0,0 +1,25 @@
+    public Map<Character, Integer> calculateUniqueCharactersAndTheirQuantity(String sentence) {
+
+        if (sentence == null) {
+            throw new NullPointerException("parameter can`t be Null");
```
please never use `NullPointerException` in your code.
we have more suitable exception for illegal parameters `IllegalArgumentException`
also please more context to message `parameter` is too generic

ALeksey5678 replied:
Done

saber0008 commented on src/main/java/com/aleksey5678/uniquecharacterscounter/stringformatter/MapToStringFormatter.java:
```diff
@@ -0,0 +1,19 @@
+    public String formattingMapToString(Map<Character, Integer> quantityOfUniqueCharactersByCharacter) {
+
+        if (quantityOfUniqueCharactersByCharacter == null) {
+            throw new NullPointerException("parameter can`t be Null");
```
please use `IllegalArgumentException`

ALeksey5678 replied:
Done

saber0008 commented on src/main/java/com/aleksey5678/uniquecharacterscounter/stringformatter/MapToStringFormatter.java:
```diff
@@ -0,0 +1,19 @@
+public class MapToStringFormatter {
+    private static final String FORMATTING_PATTERN = "%s - %s";
+
+    public String formattingMapToString(Map<Character, Integer> quantityOfUniqueCharactersByCharacter) {
```
`format*`

ALeksey5678 replied:
Done

saber0008 commented on src/test/java/com/aleksey5678/uniquecharacterscounter/characterscounter/UniqueCharactersCounterTest.java:
```diff
@@ -0,0 +1,37 @@
+    void shouldReturnNullPointerExceptionIfParameterIsNull() {
+        String message = assertThrows(NullPointerException.class,
+                () -> uniqueCharactersCounter.calculateUniqueCharactersAndTheirQuantity(null)).getMessage();
+        String expectedMessageIfParameterIsNull = "parameter can`t be Null";
```
please put expected variables on top of method

ALeksey5678 replied:
Done

saber0008 commented on src/test/java/com/aleksey5678/uniquecharacterscounter/stringformatter/MapToStringFormatterTest.java:
```diff
@@ -0,0 +1,34 @@
+import static org.junit.jupiter.api.Assertions.assertEquals;
+import static org.junit.jupiter.api.Assertions.assertThrows;
+
+class MapToStringFormatterTest {
```
please also add test case for empty string `""`

ALeksey5678 replied:
Done

ALeksey5678 committed:
rename variables in UniqueCharactersCounter;
Add number of invocations(times())to verify method in tests
NullPointerException changed to IllegalArgumentException,add more context to message for exception throwing,the same changes in tests as well;

ALeksey5678 committed:
rename formattingMapToString to formatMapToString;
put expected variables on top of method and rename test methods for exceptions;

ALeksey5678 committed:
Add test case for empty string
rename test case method from shouldModifyQuantityOfSymbolRepetitions to shouldFormatQuantityOfSymbolRepetitions

saber0008 commented on src/main/java/com/aleksey5678/uniquecharacterscounter/characterscounter/UniqueCharactersCounter.java:
```diff
@@ -0,0 +1,25 @@
+
+        for (char character : sentence.toCharArray()) {
+            numberOfCharacterRepetitionsByUniqueCharacter.computeIfPresent(character,
+                    (symbol, integer) -> integer + ONE);
```
Name `integer` is too generic.
Lets name it `quantity`

ALeksey5678 replied:
Done

saber0008 commented on src/test/java/com/aleksey5678/uniquecharacterscounter/stringformatter/MapToStringFormatterTest.java:
```diff
@@ -0,0 +1,45 @@
+    }
+    @Test
+    void shouldReturnEmptyStringIfMapIsEmpty(){
+        String expectedMessage="";
```
formatting

ALeksey5678 replied:
Done

saber0008 commented on src/main/java/com/aleksey5678/uniquecharacterscounter/stringformatter/MapToStringFormatter.java:
```diff
@@ -0,0 +1,19 @@
+    public String formatMapToString(Map<Character, Integer> quantityOfUniqueCharactersByCharacter) {
+
+        if (quantityOfUniqueCharactersByCharacter == null) {
+            throw new IllegalArgumentException ("quantityOfUniqueCharactersByCharacter can`t be Null");
```
formatting

ALeksey5678 replied:
Done

ALeksey5678 committed:
rename variable from integer to quantity;
some formatting in MapToStringFormatter/MapToStringFormatterTest
